packages/cryptoplease/lib/features/popular_tokens/src/repository.dart

```diff
@@ -0,0 +1,48 @@
+import 'package:dfunc/dfunc.dart';
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:injectable/injectable.dart';
+
+import '../../../config.dart';
+import '../../../core/tokens/token.dart';
+import 'data/coingecko_client.dart';
+
+@injectable
+class MarketDetailsRepository {
+  MarketDetailsRepository({
+    required MarketsCoingeckoClient coingeckoClient,
+  }) : _coingeckoClient = coingeckoClient;
+
+  final MarketsCoingeckoClient _coingeckoClient;
+
+  AsyncResult<IMap<Token, double>> getTopMarketTokens({
+    required String currency,
+    required int noOfTokens,
+  }) async =>
+      _coingeckoClient
+          .getMarketTokens(
+            MarketsRequestDto(
+              vsCurrency: currency,
+              order: 'market_cap_desc',
+              perPage: noOfTokens,
+            ),
+          )
+          .toEither()
+          .mapAsync(
+            (responses) => IMap({
+              for (var e in responses) e.fromCoingecko(): e.currentPrice ?? 0
+            }),
+          );
+}
+
+extension on MarketsResponseDto {
+  Token fromCoingecko() => Token(
```

Comment:
@justinenerio 

This is enough for displaying token details (since this information comes from coingecko as well), but for buy/sell we need to know real token addresses, coming from `TokenList`.

It needs to be addressed before implementing buy/sell flow.